core/src/main/scala/besom/internal/ProtobufUtil.scala

```diff
@@ -1,42 +1,186 @@
+      case Value(Kind.StringValue(urn), _) => Right(ResourceValue(urn.asValue, None))
+      case v => Left(Exception(s"Unexpected urn value type: ${v.valueType}"))
 
-  extension (v: Vector[Value]) def asValue: Value = Value(Kind.ListValue(ListValue(v)))
+  extension (r: ResourceValue)
+    def asValue: Value = {
+      Map(
+        SpecialSig.Key -> SpecialSig.ResourceSig.asValue,
+        ResourceUrnName -> r.urn
+      ) ++ r.id.map(i => Map(ResourceIdName -> i)).getOrElse(Map.empty)
+    }.asValue
 
-  val Null: Value = Value(Kind.NullValue(NullValue.NULL_VALUE))
+case class SecretValue(value: Value):
+  def isKnown: Boolean = value.kind.stringValue.map(_ == UnknownStringValue).getOrElse(false)
+
+object SecretValue:
+  def unknown: SecretValue = SecretValue(UnknownStringValue.asValue)
+  def unapply(value: Value): Option[SecretValue] =
+    value.withSpecialSignature { case (struct, SpecialSig.SecretSig) =>
+      val value = struct.fields.getOrElse(ValueName, Null)
+      SecretValue(value)
+    }
+
+  extension (s: SecretValue)
+    def asValue: Value = Map(
+      SpecialSig.Key -> SpecialSig.SecretSig.asValue,
+      ValueName -> s.value
+    ).asValue
+
+case class OutputValue(value: Option[Value], isSecret: Boolean, dependencies: List[URN]):
+  def isKnown: Boolean = value.isDefined
+
+object OutputValue:
+  def known(value: Value, isSecret: Boolean, dependencies: List[URN]): OutputValue = OutputValue(Some(value), isSecret, dependencies)
+  def unknown(isSecret: Boolean, dependencies: List[URN]): OutputValue             = OutputValue(None, isSecret, dependencies)
+  def unapply(value: Value): Option[OutputValue] =
+    value.withSpecialSignature { case (struct, SpecialSig.OutputSig) =>
+      val value            = struct.fields.get(ValueName)
+      val isSecret         = struct.fields.get(SecretName).flatMap(_.kind.boolValue).getOrElse(false)
+      val dependencyValues = struct.fields.get(DependenciesName).flatMap(_.kind.listValue).map(_.values.toList).getOrElse(Nil)
+      val dependencies     = dependencyValues.flatMap(_.kind.stringValue).flatMap(s => URN.from(s).toOption)
+      OutputValue(value, isSecret, dependencies)
+    }
+
+  extension (o: OutputValue)
+    // noinspection ScalaUnusedSymbol
+    private[internal] def asRawValueOrUnknown: Value = o.value.getOrElse(UnknownStringValue.asValue)
+    def asValue: Value = {
+      Map(
+        SpecialSig.Key -> SpecialSig.OutputSig.asValue
+      ) ++ (if o.isKnown then Map(ValueName -> o.value.get) else Map.empty)
+        ++ (if o.isSecret then Map(SecretName -> true.asValue) else Map.empty)
+        ++ (if o.dependencies.nonEmpty
+            then Map(DependenciesName -> o.dependencies.map(_.asString.asValue).asValue)
+            else Map.empty)
+    }.asValue
+
+case class Metadata(
+  known: Boolean,
+  secret: Boolean,
+  empty: Boolean,
+  dependencies: List[URN]
+):
```

Comment:
sounds like a good idea, but as an optional followup, mainly in the interest of prioritization, since I don't see an immediate value from refactoring it right now, if we hit any code scalability issues with this abstraction we can always refactor then